torch/csrc/autograd/engine.cpp

```diff
@@ -742,44 +752,134 @@ auto Engine::execute(const edge_list& roots,
+  std::unique_lock<std::mutex> lock(print_mutex_);
+  std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
+  // LOG(ERROR) << "Engine::execute_until_ready_queue_empty on graphTask: " << graph_task;
+  while(!graph_task_rq->empty()) {
+    auto task = graph_task_rq->pop();
+    execute_node_task(task);
+    LOG(INFO)<<"executed task fn: " << task->fn_->name()<< " instance: "<< task->fn_
+      << ", for graph_task: " << graph_task << " with outstanding task num:" << graph_task->outstanding_tasks_.load();
+  }
+  // Check if we've completed execution.
+  if (graph_task_completed(graph_task)) {
+    // We don't need to explicitly notify the owner thread, since
+    // 'mark_graph_task_completed' would mark the Future as completed and this
+    // would notify the owner thread that the task has been completed.
+    mark_graph_task_completed(graph_task);
+    LOG(INFO) <<"finished graph_task: " << graph_task;
+  } else {
+    // schedule a continuation
+    at::launch([this, graph_task]() {
+        execute_until_ready_queue_empty(graph_task);
+    });
+    LOG(INFO) << "scheduled a continuation on graph_task: "  << graph_task
+      << " with outstanding task num:" << graph_task->outstanding_tasks_.load();
+  }
+
 }
 
 std::shared_ptr<FutureVariableList> Engine::execute_with_graph_task(
     const std::shared_ptr<GraphTask>& graph_task,
-    std::shared_ptr<Node> graph_root) {
-  initialize_threads_pool();
+    std::shared_ptr<Node> graph_root,
+    bool async_mode) {
+  initialize_device_threads_pool();
   // Lock mutex for GraphTask.
   std::unique_lock<std::mutex> lock(graph_task->mutex_);
 
-  ready_queue(at::kCPU).push(
+  ready_queue(graph_task, at::kCPU).push(
       NodeTask(graph_task, std::move(graph_root), InputBuffer(0)));
 
-  // Not a worker
+  // worker_device == NO_DEVICE it's a CPU thread and it's trying to drive the
+  // autograd engine with corresponding GraphTask, and its NOT a re-entrant call
   if (worker_device == NO_DEVICE) {
     // graph_task_exec_post_processing is done when the Future is marked as
     // completed in mark_graph_task_completed.
-    return graph_task->future_result_;
+
+    // We set the worker_device to CPU_DEVICE only if worker_device was previously
+    // NO_DEVICE. Setting it to CPU afterwards allow us to detect whether this is
+    // a re-entrant call or not.
+    //
+    // If worker_device = NO_DEVICE: this is NOT a re-entrant call
+    // If worker_device is other devices (i.e. CPU, CUDA): this is a re-entrant
+    //    backward call from that device.
+    set_device(CPU_DEVICE);
+
+    // set the graph_task owner to the current device
+    graph_task->owner_ = worker_device;
+
+    // The owning thread start to drive the engine execution with the GraphTask
+    // that has already been pushed to the current CPU thread's ready_queue
+    lock.unlock();
+    if (async_mode) {
+      execute_until_ready_queue_empty(graph_task);
```

Comment:
I think the semantic of this async mode is a bit unexpected:
- It runs part of the work in a sync manner
- The rest will run in async

Why not just run the whole thing in async mode and make this thread return instantly?
@pritamdamania87 would that match your expectations for the distributed engine?